stdlib/public/core/SwiftNativeNSArray.swift

```diff
@@ -129,6 +141,189 @@ extension __SwiftNativeNSArrayWithContiguousStorage: _NSArrayCore {
   internal func copy(with _: _SwiftNSZone?) -> AnyObject {
     return self
   }
+  
+  @objc(mutableCopyWithZone:)
+  internal func mutableCopy(with _: _SwiftNSZone?) -> AnyObject {
+    let arr = Array<AnyObject(_ContiguousArrayBuffer(self))
+    return _SwiftNSMutableArray(arr)
+  }
+  
+  @objc(indexOfObjectIdenticalTo:)
+  internal func index(ofObjectIdenticalTo object: AnyObject) -> Int {
+    let arr = Array<AnyObject(_ContiguousArrayBuffer(self))
+    return arr.firstIndex { $0 === object } ?? NSNotFound
+  }
+}
+
+@_fixed_layout
+@usableFromInline
+@objc internal final class _SwiftNSMutableArray :
+  _SwiftNativeNSMutableArray, _NSArrayCore
+{
+  internal var contents: [AnyObject]
+
+  internal init(_ array: [AnyObject]) {
+    contents = array
+    super.init()
+  }
+  
+  @objc internal var count: Int {
+    return contents.count
+  }
+  
+  @objc(objectAtIndexedSubscript:)
+  internal func objectAtSubscript(_ index: Int) -> AnyObject {
+    //TODO: exception instead of precondition, once that's possible
+    return contents[index]
+  }
+
+  @objc(objectAtIndex:)
+  internal func objectAt(_ index: Int) -> AnyObject {
+    //TODO: exception instead of precondition, once that's possible
+    return contents[index]
+  }
+
+  @objc internal func getObjects(
+    _ aBuffer: UnsafeMutablePointer<AnyObject>, range: _SwiftNSRange
+  ) {
+    return contents.withContiguousStorageIfAvailable { objects in
+      //TODO: exceptions instead of preconditions, once that's possible
+
+      _precondition(
+        _isValidArrayIndex(range.location, count: objects.count),
+        "Array index out of range")
+
+      _precondition(
+        _isValidArrayIndex(
+          range.location + range.length, count: objects.count),
+        "Array index out of range")
+
+      if objects.isEmpty { return }
+
+      // These objects are "returned" at +0, so treat them as pointer values to
+      // avoid retains. Copy bytes via a raw pointer to circumvent reference
+      // counting while correctly aliasing with all other pointer types.
+      UnsafeMutableRawPointer(aBuffer).copyMemory(
+        from: objects.baseAddress! + range.location,
+        byteCount: range.length * MemoryLayout<AnyObject>.stride)
+    }!
+  }
+
+  @objc(countByEnumeratingWithState:objects:count:)
+  internal func countByEnumerating(
+    with state: UnsafeMutablePointer<_SwiftNSFastEnumerationState>,
+    objects: UnsafeMutablePointer<AnyObject>?, count: Int
+  ) -> Int {
+    var enumerationState = state.pointee
+
+    if enumerationState.state != 0 {
+      return 0
+    }
+
+    return contents.withContiguousStorageIfAvailable {
+      objects in
+      enumerationState.mutationsPtr = _fastEnumerationStorageMutationsPtr
+      enumerationState.itemsPtr =
+        AutoreleasingUnsafeMutablePointer(objects.baseAddress)
```

Comment:
This is very much violating the spirit of `withContiguousStorageIfAvailable`, which still says the address isn't valid past the end of the call.

Reply:
All the bits here and in the comment above are copy-pasted from the other subclass, so I guess we should look into fixing both :(

Reply:
We actually *do* have to guard against mutations too, right?

Reply:
They should trigger CoW, but I should think about whether that's what we want

Reply:
They won't trigger CoW because NSFastEnumeration doesn't actually hold a reference to the array buffer itself.

Reply:
I forget how `_fastEnumerationStorageMutationsPtr` works, though, it may be sufficient.

Reply:
Arguably the whole point is "modifying an array while it's being enumerated is disallowed", but it would be good if that meant it crashed deterministically rather than just silently accessing garbage memory. `_fastEnumerationStorageMutationsPtr` might be sufficient for that.